## Imports


In [1]:
import random
import re
import time

from nltk import word_tokenize
from nltk.corpus import stopwords
import numpy as np

from nltk.book import (
    text1,
    text2,
    text3,
    text4,
    text5,
    text6,
    text7,
    text8,
    text9,
    Text,
)

*** Introductory Examples for the NLTK Book ***
Loading text1, ..., text9 and sent1, ..., sent9
Type the name of the text or sentence to view it.
Type: 'texts()' or 'sents()' to list the materials.
text1: Moby Dick by Herman Melville 1851
text2: Sense and Sensibility by Jane Austen 1811
text3: The Book of Genesis
text4: Inaugural Address Corpus
text5: Chat Corpus
text6: Monty Python and the Holy Grail
text7: Wall Street Journal
text8: Personals Corpus
text9: The Man Who Was Thursday by G . K . Chesterton 1908


## Question 1 A

Generates a list of 2000 random words from one of the ntlk's book corpuses, using a random seed (of any method). Which set of text to be used should also be determined randomly from:

- text1: Moby Dick by Herman Melville 1851
- text2: Sense and Sensibility by Jane Austen 1811
- text3: The Book of Genesis
- text4: Inaugural Address Corpus
- text5: Chat Corpus
- text6: Monty Python and the Holy Grail
- text7: Wall Street Journal
- text8: Personals Corpus


In [2]:
all_texts = [text1, text2, text3, text4, text5, text6, text7, text8, text9]


def select_random_text(all_texts, num_words):
    text: Text = random.choice(all_texts)
    print(f"Using {text}")

    text_data = text.generate(num_words, random_seed=time.time())
    return text_data


text = select_random_text(all_texts=all_texts, num_words=2000)

Using <Text: Wall Street Journal>


Building ngram index...


, our rating was maintained *-1 , analysts said 0 it is scheduled *-1
to win reinstatement , and some stations say 0 *T*-3 , they want *-3
to sell TXO 's oil and gas and chemicals . easing from September when
8.1 % anticipated a purchase *T*-1 . on preventative medicine .
*T*-140 may have single-handedly turned around ratings at NBC since
its debut this fall despite the bald-faced nature of her , while
International Paper fell 7\/8 to 10 on volume of junk-bond financings
shrinks . magazine 's editor in chief . White contributed to its
foreign counterparts has helped *-1 much by the House Energy and
Commerce departments . they believe 0 * to offer a bargain . '' deals
also improve Japanese access to the apparent slowing of the new shares
has been seeking a new contract and on its common stock , which *T*-12
began in 1987 *-2 to reprint such material , '' a poignant piece of a
vitriolic U.S. response to the 30-point limit indicated that the group
's recent solid performance against the r

## Question 2

Creates a list of word tokens from this corpus


In [3]:
def tokenize_text(text):
    return np.array(word_tokenize(text))


word_tokens = tokenize_text(text)

## Question 3

Removes non-words from the list of tokens. Assumes numbers are not to be considerd words


In [4]:
def remove_non_words(word_tokens):
    tokens_without_non_words = []
    for word in word_tokens:
        if not re.search(r"[a-zA-z]", word):
            print(f"Removing a non-word: {word}")
            continue
        tokens_without_non_words.append(word)
    return np.array(tokens_without_non_words)


tokens_without_non_words = remove_non_words(word_tokens)

Removing a non-word: ,
Removing a non-word: *
Removing a non-word: -1
Removing a non-word: ,
Removing a non-word: 0
Removing a non-word: *
Removing a non-word: -1
Removing a non-word: ,
Removing a non-word: 0
Removing a non-word: *
Removing a non-word: *
Removing a non-word: -3
Removing a non-word: ,
Removing a non-word: *
Removing a non-word: -3
Removing a non-word: .
Removing a non-word: 8.1
Removing a non-word: %
Removing a non-word: *
Removing a non-word: *
Removing a non-word: -1
Removing a non-word: .
Removing a non-word: .
Removing a non-word: *
Removing a non-word: *
Removing a non-word: -140
Removing a non-word: ,
Removing a non-word: 10
Removing a non-word: .
Removing a non-word: .
Removing a non-word: *
Removing a non-word: -1
Removing a non-word: .
Removing a non-word: 0
Removing a non-word: *
Removing a non-word: .
Removing a non-word: ,
Removing a non-word: *
Removing a non-word: *
Removing a non-word: -12
Removing a non-word: 1987
Removing a non-word: *
Removing a non-wo

## Question 4

Removes stop-words using the ntlk's English stop-word list


In [5]:
stop_words = stopwords.words("english")


def remove_stop_words(tokens_without_non_words, stop_words):
    tokens_without_stop_words = []
    for word in tokens_without_non_words:
        if word in stop_words:
            print(f"Removing a stopword: {word}")
            continue
        tokens_without_stop_words.append(word)
    return np.array(tokens_without_stop_words)


tokens_without_non_and_stop_words = remove_stop_words(
    tokens_without_non_words, stop_words
)

Removing a stopword: our
Removing a stopword: was
Removing a stopword: it
Removing a stopword: is
Removing a stopword: to
Removing a stopword: and
Removing a stopword: some
Removing a stopword: they
Removing a stopword: to
Removing a stopword: and
Removing a stopword: and
Removing a stopword: from
Removing a stopword: when
Removing a stopword: a
Removing a stopword: on
Removing a stopword: have
Removing a stopword: at
Removing a stopword: its
Removing a stopword: this
Removing a stopword: the
Removing a stopword: of
Removing a stopword: her
Removing a stopword: while
Removing a stopword: to
Removing a stopword: on
Removing a stopword: of
Removing a stopword: in
Removing a stopword: to
Removing a stopword: its
Removing a stopword: has
Removing a stopword: by
Removing a stopword: the
Removing a stopword: and
Removing a stopword: they
Removing a stopword: to
Removing a stopword: a
Removing a stopword: to
Removing a stopword: the
Removing a stopword: of
Removing a stopword: the
Removing a 

## Question 5

Prints out a word that appears most often in the corpus as well its frequency


In [6]:
def most_frequent(word_tokens):
    unique, counts = np.unique(word_tokens, return_counts=True)
    index_of_most_frequent = np.argmax(counts)
    most_frequent_word = unique[index_of_most_frequent]
    count_of_most_frequent = counts[index_of_most_frequent]

    print(
        f"'{most_frequent_word}' appeared most often ( {count_of_most_frequent} times )"
    )


most_frequent(tokens_without_non_and_stop_words)

'T' appeared most often ( 33 times )


## Question 6

Again, run the entire code multiple times, and observe that the output changes (almost) every time:


In [8]:
def repeat(repeats):
    for i in range(1, repeats + 1):
        print(f"-------------------- Try {i} --------------------")
        text = select_random_text(all_texts=all_texts, num_words=2000)
        word_tokens = tokenize_text(text)
        tokens_without_non_words = remove_non_words(word_tokens)
        tokens_without_non_and_stop_words = remove_stop_words(
            tokens_without_non_words, stop_words
        )
        most_frequent(tokens_without_non_and_stop_words)
        print(f"-----------------------------------------------------\n\n")


repeat(3)

-------------------- Try 1 --------------------
Using <Text: The Man Who Was Thursday by G . K . Chesterton 1908>


Building ngram index...


have guessed that he had imagined that all was lost ; but it either
was not a curate ( loud and cheerful voice . " What the devil ; he can
go away to right or wrong , they saw the frightful figure that might
be smoked or knives drawn . , " except that I cannot help thinking
that it is positive , as he seated himself at the gallop , had stood
up in his company . " me , though it destroyed nations ; but each man
of very massive common sense a little faster , so that he ought to be
dragging out like yellow rags of rank grass , his handsome face brazen
in the round , black cigar , bought in Soho for twopence , stood seven
great chairs , the Marquis have insulted your aunt just now . that
barrel - organ begin to play , at a Sunday teashop , turned in his
teeth ; he was the same hansom cab from Victoria ; but he cannot have
carried the world is not twice one ; two is not dead in a sudden
effort to rise for him , as he said . " It is indeed a new planet . AT
first the large face of an ox . an

Building ngram index...


heard by her , without whose interest it was beyond her first
conversation with our books and other refreshments ; and to amuse her
a moment to her sister , the eldest of them they raised a much higher
situation than the immediate distress of hearing of her effects at
Norland about a week should be married ,-- and come while the same
thought ; and should it be ? John , and finding no good comes of it
with earnestness , he is , that if Edward might have determined him to
be surrounded with difficulties on every line was an obligation . "
but the very person -- for the huswifes she had not said the day , at
least -- my brother and sister will be a woman of higher rank and
larger fortune ;-- and her whole heart to you , and writing as fast as
could not get rid of her daughters , both of them in friendship , or
deliver it in his style of a fever ?" her into a reverie of this
cottage that will open his character , was only such as a measure
which would be relieved from the first , indeed ! 